ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -47,6 +48,359 @@ struct PendingRenameOperation
     thread_t thread;
 };
 
+class NullTracer
+{
+    NullTracer() = delete;
+    NullTracer(const NullTracer&) = delete;
+    NullTracer& operator=(const NullTracer&) = delete;
+public:
+    explicit NullTracer(kauth_action_t vnodeAction, vnode_t vnode)
+    {}
+    explicit NullTracer(kauth_action_t fileopAction, uintptr_t arg0, uintptr_t arg1, uintptr_t arg2)
+    {}
+    
+    void Flush()
+    {
+    }
+    void Printf(const char* format, ...) __attribute__ ((format (printf, 2, 3)))
+    {
+    }
+    void SendProviderMessage(MessageType providerMessage)
+    {}
+    void SendProviderMessageResult(bool success)
+    {}
+    
+    void SetDeniedForCrawler()
+    {}
+
+    void SetVnodeOpResult(int result)
+    {
+    }
+};
+
+static const char* MessageTypeString(MessageType messageType)
+{
+    switch(messageType)
+    {
+    case MessageType_KtoU_EnumerateDirectory:            return "EnumerateDirectory";
+    case MessageType_KtoU_RecursivelyEnumerateDirectory: return "RecursivelyEnumerateDirectory";
+    case MessageType_KtoU_HydrateFile:                   return "HydrateFile";
+    case MessageType_KtoU_NotifyFileModified:            return "NotifyFileModified";
+    case MessageType_KtoU_NotifyFilePreDelete:           return "NotifyFilePreDelete";
+    case MessageType_KtoU_NotifyFilePreDeleteFromRename: return "NotifyFilePreDeleteFromRename";
+    case MessageType_KtoU_NotifyDirectoryPreDelete:      return "NotifyDirectoryPreDelete";
+    case MessageType_KtoU_NotifyFileCreated:             return "NotifyFileCreated";
+    case MessageType_KtoU_NotifyFileRenamed:             return "NotifyFileRenamed";
+    case MessageType_KtoU_NotifyDirectoryRenamed:        return "NotifyDirectoryRenamed";
+    case MessageType_KtoU_NotifyFileHardLinkCreated:     return "NotifyFileHardLinkCreated";
+    case MessageType_KtoU_NotifyFilePreConvertToFull:    return "NotifyFilePreConvertToFull";
+    default:                                             return "Unknown";
+    };
+}
+
+class KextLogTracer
```

Comment:
Would it be worth moving some of the logic to it's own class.  KextLogDiagnostics or similar?

Reply:
What functionality would you split out? I guess the buffer could be encapsulated.